### PR TITLE
Create log folder if not present

### LINKED ISSUE #1217: Traefik unable to create logs folder in docker swarm
### What version of Traefik are you using (`traefik version`)?
1.1.2

### What is your environment & configuration (arguments, toml...)?
We are running traefik as docker service in swarm mode.

### What did you do?
 The following docker command is used to run traefik

```
sudo docker service create \
    --name traefik \
    --replicas=1 \
    --constraint=node.role==manager \
    --publish 80:80 --publish 8080:8080 \
    --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
    --network swarm-demo-overlay-net \
    traefik:v1.1.2 \
    --docker \
    --docker.swarmmode \
    --docker.domain=abc.com \
    --docker.watch \
    --logLevel=INFO \
    --traefikLogsFile='/app/logs/traefik.log' \
    --accessLogsFile='/app/logs/access.log' \
    --web

```
### What did you expect to see?
Traefik should create the log folder '/app/logs/'  if not present.

### What did you see instead?
Traefik throws following error at stdout and then redirects the logs to stdout

```shell
time="2017-03-02T04:13:55Z" level=error msg="Error opening fileopen \"/app/logs/access.log\": no such file or directory" 
time="2017-03-02T04:13:55Z" level=error msg="Error opening fileopen \"/app/logs/traefik.log\": no such file or directory" 
time="2017-03-02T04:13:55Z" level=info msg="Traefik version v1.1.2 built on 2016-12-15_10:21:15AM" 
time="2017-03-02T04:13:55Z" level=error msg="Error opening fileopen \"/app/logs/access.log\": no such file or directory" 

```

### PR DESCRIPTION
**Pull request by tanyadegurechaff**
### Description

Fixes #1217

### CODE REVIEW

File: middlewares/logger_test.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ type logtestResponseWriter struct{}
 
 var (
 	logger                  *Logger
-	logfileName             = "traefikTestLogger.log"
+	logfileName             = "/traefik/logger/test.log"
```
Comment (timoreimann) [inline middlewares/logger_test.go]: Could you rename this to `logfileNameSuffix` since it will be concatenated with the actual temp directory.

Comment (timoreimann) [inline middlewares/logger_test.go]: Could you please also remove the global variable `logfilePath`? It suffices to live on the test function stack.

**Comment by tanyadegurechaff** [comment]
Thank you for your feedbacks @timoreimann

**Comment by timoreimann** [comment]
Apparently, log files are now created at two locations. Is that necessary? Aren't those the same log files?

**Comment by tanyadegurechaff** [comment]
I'm afraid not. One is traefik log file, the other is access log file.

File: middlewares/logger_test.go [hunk 2]
```diff
@@ -41,14 +39,21 @@ var (
 )
 
 func TestLogger(t *testing.T) {
-	if runtime.GOOS == "windows" {
-		logfilePath = filepath.Join(os.Getenv("TEMP"), logfileName)
-	} else {
-		logfilePath = filepath.Join("/tmp", logfileName)
+	tmp, err := ioutil.TempDir("", "testlogger")
+	if err != nil {
+		t.Fatal("failed to create temp dir: ", err)
```
Comment (timoreimann) [inline middlewares/logger_test.go]: Can we use `Fatalf` and `%s` just for consistency reasons?

File: middlewares/logger_test.go [hunk 3]
```diff
@@ -41,14 +39,21 @@ var (
...
+	defer os.RemoveAll(tmp)
+
+	logfilePath := filepath.Join(tmp, logfileNameSuffix)
 
 	logger = NewLogger(logfilePath)
-	defer cleanup()
+	defer logger.Close()
+
+	if _, err := os.Stat(logfilePath); err != nil {
+		t.Fatalf("logger should create %q", logfilePath)
```
Comment (timoreimann) [inline middlewares/logger_test.go]: Any particular reason we're using `%q` here as opposed to just `%s`?

File: cmd/traefik/traefik.go [hunk 4]
```diff
@@ -206,6 +207,13 @@ func run(traefikConfiguration *server.TraefikConfiguration) {
 	}
 	log.SetLevel(level)
 	if len(globalConfiguration.TraefikLogsFile) > 0 {
+		dir := filepath.Dir(globalConfiguration.TraefikLogsFile)
+
+		err := os.MkdirAll(dir, 0755)
+		if err != nil {
+			log.Errorf("Failed to create log path %s: %s", dir, err)
+		}
+
```
Comment (timoreimann) [inline cmd/traefik/traefik.go]: The file seems completely untested, but making it properly testable is likely a bigger refactoring we want to source out to a different PR.

Could you verify manually that this part of the code change works as expected please?

File: middlewares/logger_test.go [hunk 5]
```diff
@@ -41,14 +39,21 @@ var (
...
 	}
+	defer os.RemoveAll(tmp)
+
+	logfilePath := filepath.Join(tmp, logfileNameSuffix)
 
 	logger = NewLogger(logfilePath)
-	defer cleanup()
+	defer logger.Close()
+
+	if _, err := os.Stat(logfilePath); err != nil {
```
Comment (timoreimann) [inline middlewares/logger_test.go]: I think you need [os.IsNotExist](http://stackoverflow.com/a/12518877/420061), don't you?

File: cmd/traefik/traefik.go (hunk 4 above)
Comment (tanyadegurechaff) [inline cmd/traefik/traefik.go]: Sure.

**Comment by timoreimann** [comment]
CI flaked out -- retriggered.

**Comment by timoreimann** [comment]
@emilevauge @ldez looks simple enough to still get into 1.3, WDYT?

**Comment by emilevauge** [comment]
@timoreimann why not, my biggest concern is that this PR https://github.com/containous/traefik/pull/1485 is a big refactor of the logger and we will probably not be able to merge the current one once #1485 is merged. WDYT @tanyadegurechaff ?

**Comment by timoreimann** [comment]
@emilevauge this change is rather smallish. My assumption is that it should be easy for #1485 to adjust / fix a smallish merge commit.

@rjshep, any thoughts?

**Comment by timoreimann** [comment]
@emilevauge put differently, I'd rather have the big change pay an integration effort for a small change than the other way around.

Or is it that #1485 does things so fundamentally different as far as this PR is concerned that we will run into bigger problems I don't see right now?

**Comment by ldez** [comment]
Due to SemaphoreCI, I close and reopen the PR.

**Pull request merged**